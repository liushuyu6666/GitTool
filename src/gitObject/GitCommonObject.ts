// export interface GitCommonObjectInterface {
//   hash: string;

//   prefix: string;

//   suffix: string;

//   objectLoc: string;

//   dividedDecryptedBuffer: Buffer[];

//   header: string;

//   size: number;
// }