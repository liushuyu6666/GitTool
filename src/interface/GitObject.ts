// export interface GitObject {
//   hash: string;

//   prefix: string;

//   suffix: string;

//   /* absolute location of this object */
//   objectLoc: string;

//   // TODO: should be enum
//   type: string;

//   /* 
//     This is the Buffer size of the content;
//     Buffer.length != string.length if there are 
//     characters whose value larger than 127
//   */
//   size: number;
// }