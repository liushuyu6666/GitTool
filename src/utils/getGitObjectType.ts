export enum GitObjectType {
  BLOB = 'blob',
  TREE = 'tree',
  COMMIT = 'commit',
  TAG = 'tag',
  BLOB_DELTA = 'blob_delta',
  TREE_DELTA = 'tree_delta',
  COMMIT_DELTA = 'commit_delta',
  TAG_DELTA = 'tag_delta',
  OFS_DELTA = 'ofs_delta',
  REF_DELTA = 'ref_delta',
  UNDEFINED = 'undefined',
}

export const GitPackObjectType: Record<number, GitObjectType> = {
  1: GitObjectType.COMMIT_DELTA,
  2: GitObjectType.TREE_DELTA,
  3: GitObjectType.BLOB_DELTA,
  4: GitObjectType.TAG_DELTA,
  5: GitObjectType.OFS_DELTA,
  6: GitObjectType.REF_DELTA,
};

export default function (type: string | number): GitObjectType {
  if (typeof type === 'string') {
    switch (type) {
      case GitObjectType.BLOB:
        return GitObjectType.BLOB;
      case GitObjectType.TREE:
        return GitObjectType.TREE;
      case GitObjectType.COMMIT:
        return GitObjectType.COMMIT;
      case GitObjectType.TAG:
        return GitObjectType.TAG;
      case GitObjectType.OFS_DELTA:
        return GitObjectType.OFS_DELTA;
      case GitObjectType.REF_DELTA:
        return GitObjectType.REF_DELTA;
      default:
        throw new Error(`Can\'t find the GitObjectType: ${type}`);
    }
  } else if (typeof type === 'number') {
    // if it is number, then it should come from pack file
    return GitPackObjectType[type];
  } else {
    throw new Error(`getGitObjectType error!`);
  }
}

const originalType: Array<GitObjectType> = [
  GitObjectType.BLOB,
  GitObjectType.TREE,
  GitObjectType.COMMIT,
  GitObjectType.TAG,
];

const originalDeltaType: Array<GitObjectType> = [
  GitObjectType.BLOB_DELTA,
  GitObjectType.TREE_DELTA,
  GitObjectType.COMMIT_DELTA,
  GitObjectType.TAG_DELTA,
]

const deltaType: Array<GitObjectType> = [
  GitObjectType.OFS_DELTA,
  GitObjectType.REF_DELTA,
];

export function isOriginalObject(type: GitObjectType): boolean {
  return originalType.includes(type);
}

export function isOriginalDeltaObject(type: GitObjectType): boolean {
  return originalDeltaType.includes(type);
}

export function isDeltaObject(type: GitObjectType): boolean {
  return deltaType.includes(type);
}
