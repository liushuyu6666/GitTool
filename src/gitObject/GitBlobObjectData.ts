import { parseBlobObjectContent } from '../utils/gitObjectParse/blob';
import { GitObjectType } from './GitObject';

export interface GitBlobObjectInterface {
  type: GitObjectType.BLOB;

  content: string;
}

export class GitBlobObjectData implements GitBlobObjectInterface {
  type: GitObjectType.BLOB;

  content: string;

  constructor(dividedDecryptedBuffer: Buffer[]) {
    this.type = GitObjectType.BLOB;
    this.content = parseBlobObjectContent(dividedDecryptedBuffer);
  }
}
