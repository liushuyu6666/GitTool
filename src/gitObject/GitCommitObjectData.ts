import { CommitObjectInfo, parseCommitObjectContent } from '../utils/gitObjectParse/commit';
import { GitObjectType } from './GitObject';

export interface GitCommitObjectInterface {
  type: GitObjectType.COMMIT;

  commitData: CommitObjectInfo;
}

export class GitCommitObjectData implements GitCommitObjectInterface {
  type: GitObjectType.COMMIT;

  commitData: CommitObjectInfo;

  constructor(body: Buffer) {
    this.type = GitObjectType.COMMIT;
    this.commitData = parseCommitObjectContent(body);
  }
}
