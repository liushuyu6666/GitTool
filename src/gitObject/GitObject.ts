import { GitBlobObjectData } from './GitBlobObjectData';
import { GitTreeObjectData } from './GitTreeObjectData';
import { inflateSync } from 'zlib';
import fs from 'fs';
import { GitCommitObjectData } from './GitCommitObjectData';
import {
  GitObjectType,
  isOriginalObject,
  isPackDeltaObject,
  isPackOriginalObject,
} from '../utils/getGitObjectType';
import { GitRefDeltaObjectData } from './GitRefDeltaObjectData';

export enum Mode {
  normal = '100644',
  tree = '040000',
}

export interface GitObjectInterface {
  hash: string;

  type: GitObjectType;

  size: number;

  data:
    | undefined
    | GitBlobObjectData
    | GitTreeObjectData
    | GitCommitObjectData
    | GitRefDeltaObjectData;

  differentiating(): void;
}

export interface GitObjectInput {
  hash: string;

  type: GitObjectType;

  size: number;

  filePath: string;

  bodyOffsetStartIndex: number;

  bodyOffsetEndIndex: number;
}

export class GitObject implements GitObjectInterface {
  hash: string;

  prefix: string;

  suffix: string;

  type: GitObjectType;

  size: number;

  filePath: string;

  bodyOffsetStartIndex: number;

  bodyOffsetEndIndex: number;

  data:
    | undefined
    | GitBlobObjectData
    | GitTreeObjectData
    | GitCommitObjectData
    | GitRefDeltaObjectData;

  constructor({
    hash,
    type,
    size,
    filePath,
    bodyOffsetStartIndex,
    bodyOffsetEndIndex,
  }: GitObjectInput) {
    this.hash = hash ?? '';

    this.prefix = this.hash.slice(0, 2);

    this.suffix = this.hash.slice(2, this.hash.length);

    this.type = type;

    this.size = size;

    this.filePath = filePath;

    this.bodyOffsetStartIndex = bodyOffsetStartIndex;

    this.bodyOffsetEndIndex = bodyOffsetEndIndex;

    this.data = undefined;
  }

  differentiating() {
    const type = this.type;
    const content = fs.readFileSync(this.filePath);

    if (isOriginalObject(type)) {
      const decryptedBuf = inflateSync(content);
      const body = decryptedBuf.subarray(
        this.bodyOffsetStartIndex,
        this.bodyOffsetEndIndex,
      );
      switch (type) {
        case GitObjectType.BLOB:
          this.data = new GitBlobObjectData(body);
          break;
        case GitObjectType.TREE:
          this.data = new GitTreeObjectData(body, this.hash);
          break;
        case GitObjectType.COMMIT:
          this.data = new GitCommitObjectData(body);
          break;
        default:
          throw new Error(
            `${this.hash} can\'t be differentiated in the GitObject.`,
          );
      }
    } else if (isPackOriginalObject(type)) {
      const body = content.subarray(
        this.bodyOffsetStartIndex,
        this.bodyOffsetEndIndex,
      );
      const decryptedBuf = inflateSync(body);
      switch (type) {
        case GitObjectType.BLOB_DELTA:
          this.data = new GitBlobObjectData(decryptedBuf);
          break;
        case GitObjectType.TREE_DELTA:
          this.data = new GitTreeObjectData(decryptedBuf, this.hash);
          break;
        case GitObjectType.COMMIT_DELTA:
          this.data = new GitCommitObjectData(decryptedBuf);
          break;
        default:
          throw new Error(
            `${this.hash} can\'t be differentiated in the GitObject.`,
          );
      }
    } else if (isPackDeltaObject(type)) {
      const body = content.subarray(
        this.bodyOffsetStartIndex,
        this.bodyOffsetEndIndex,
      );
      console.log(body.toString('hex'));
      switch (type) {
        case GitObjectType.REF_DELTA:
          this.data = new GitRefDeltaObjectData(body);
          break;
      }
    }
  }
}
