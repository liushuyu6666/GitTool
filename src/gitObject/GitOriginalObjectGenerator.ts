import fs from 'fs';
import path from 'path';
import { inflateSync } from 'zlib';
import getGitObjectType from '../utils/getGitObjectType';
import splitHeaderAndBody from '../utils/splitHeaderAndBody';
import { GitObject } from './GitObject';

export interface GitOriginalObjectGeneratorInterface {
  filePath: string;

  generateGitOriginalObject(): GitObject
}

/**
 * 
 * @param filePath The file path of the original object, always likes this `${rooDir}/.git/objects/e3/e1dd343baa984e5d704c54404a80de05a5008e`
 */
export class GitOriginalObjectGenerator implements GitOriginalObjectGeneratorInterface {
  filePath: string;

  constructor(filePath: string) {
    this.filePath = filePath;
  }

  generateGitOriginalObject(): GitObject {
    const filePaths: string[] = this.filePath.split(path.sep);
    const suffix = filePaths.pop();
    const prefix = filePaths.pop();
    if (!suffix || !prefix) {
      throw new Error(`GitOriginalObjectGenerator: can\'t get prefix or suffix from the file path ${this.filePath}`);
    }
    const hash = prefix + suffix;

    const decryptedBuf = inflateSync(fs.readFileSync(this.filePath));
    const [header, _] = splitHeaderAndBody(decryptedBuf);
    const headerStr = header.toString();
    const size = parseInt(headerStr.split(' ')[1]);
    const type = getGitObjectType(headerStr.split(' ')[0])

    const bodyOffsetStartIndex = header.length + 1;
    const bodyOffsetEndIndex = decryptedBuf.length;

    return new GitObject({
      hash,
      type,
      size,
      filePath: this.filePath,
      bodyOffsetStartIndex,
      bodyOffsetEndIndex
    })
  }
}