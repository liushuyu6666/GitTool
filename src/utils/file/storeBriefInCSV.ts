import fs from 'fs';
import { GitObject } from '../../object/gitObject/GitObject';

const newLine = '\r\n';

export default function (filePath: string, gitObject: GitObject) {
  const fields = Object.keys(gitObject).filter((field) => field !== 'data');

  // Create a new CSV file with fields
  if (!fs.existsSync(filePath)) {
    console.log(`${filePath} is creating`);
    fs.writeFileSync(filePath, fields.join(',') + newLine);
  }

  const csvData = Object.values(gitObject).join(',');
  fs.appendFileSync(filePath, csvData + newLine);
}
