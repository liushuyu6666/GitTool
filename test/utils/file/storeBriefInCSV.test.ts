import fs from "fs";
import { GitObject } from "../../../src/object/gitObject/GitObject";
import storeBriefInCSV from "../../../src/utils/file/storeBriefInCSV";
import { GitObjectType } from "../../../src/utils/getGitObjectType";

describe('Test the storeBriefInCSV function', () => {
  const brief1 = new GitObject({
    hash: 'test111111111111',
    type: GitObjectType.BLOB,
    size: 1,
    filePath: 'test_object',
    bodyOffsetStartIndex: 34,
    bodyOffsetEndIndex: 100
  });

  const brief2 = new GitObject({
    hash: 'test222222222222',
    type: GitObjectType.BLOB,
    size: 2,
    filePath: 'test_object2',
    bodyOffsetStartIndex: 13,
    bodyOffsetEndIndex: 100
  });

  test('Saving the first record in the csv file', () => {
  
    const file = 'test/utils/file/brief1.csv';

    // delete the file if exists
    if (fs.existsSync(file)) {
      fs.unlinkSync(file);
    }
    
    storeBriefInCSV(file, brief1);
    const csv = fs.readFileSync(file, 'utf-8');
    expect(csv).toMatchSnapshot();
  });

  test('Saving the second record in the csv file', () => {
  
    const file = 'test/utils/file/brief2.csv';

    // delete the file if exists
    if (fs.existsSync(file)) {
      fs.unlinkSync(file);
    }

    storeBriefInCSV(file, brief1);
    storeBriefInCSV(file, brief2);
    const csv = fs.readFileSync(file, 'utf-8');
    expect(csv).toMatchSnapshot();
  });
})