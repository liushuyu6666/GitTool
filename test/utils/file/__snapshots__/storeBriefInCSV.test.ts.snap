// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the storeBriefInCSV function Saving the first record in the csv file 1`] = `
"hash,prefix,suffix,type,size,filePath,bodyOffsetStartIndex,bodyOffsetEndIndex
test111111111111,te,st111111111111,blob,1,test_object,34,100,
"
`;

exports[`Test the storeBriefInCSV function Saving the second record in the csv file 1`] = `
"hash,prefix,suffix,type,size,filePath,bodyOffsetStartIndex,bodyOffsetEndIndex
test111111111111,te,st111111111111,blob,1,test_object,34,100,
test222222222222,te,st222222222222,blob,2,test_object2,13,100,
"
`;
